Detectors/DCS/include/DetectorsDCS/DCSProcessor.h

```diff
@@ -0,0 +1,299 @@
+  //omp_set_num_threads(mNThreads);
+  //#pragma omp parallel for schedule(dynamic)
+  //#endif
+  for (size_t i = 0; i != array.size(); ++i) {
+    auto it = findAndCheckAlias(array[i], type, map);
+    if (it == map.end()) {
+      if (!mIsDelta) {
+        LOG(ERROR) << "Element " << array[i] << " not found " << std::endl;
+      }
+      continue;
+    }
+    found++;
+    std::pair<DPID, DPVAL> pairIt = *it;
+    checkFlagsAndFill(pairIt, latestTimeStamp[i], destmap);
+    if (type == RAW_CHAR) {
+      processCharDP(array[i]);
+    } else if (type == RAW_INT) {
+      processIntDP(array[i]);
+    } else if (type == RAW_DOUBLE) {
+      processDoubleDP(array[i]);
+    } else if (type == RAW_UINT) {
+      processUIntDP(array[i]);
+    } else if (type == RAW_BOOL) {
+      processBoolDP(array[i]);
+    } else if (type == RAW_STRING) {
+      processStringDP(array[i]);
+    } else if (type == RAW_TIME) {
+      processTimeDP(array[i]);
+    } else if (type == RAW_BINARY) {
+      processBinaryDP(array[i]);
+    }
+    // todo: better to move the "found++" after the process, in case it fails?
+  }
+  return found;
+}
+
+//______________________________________________________________________
+
+template <typename T>
+void DCSProcessor::checkFlagsAndFill(const std::pair<DPID, DPVAL>& dpcom, int64_t& latestTimeStamp, std::unordered_map<DPID, T>& destmap)
+{
+
+  // check the flags for the upcoming data, and if ok, fill the accumulator
+
+  auto& dpid = dpcom.first;
+  auto& val = dpcom.second;
+  auto flags = val.get_flags();
+  if (processFlag(flags, dpid.get_alias()) == 0) {
+    auto etime = val.get_epoch_time();
+    // fill only if new value has a timestamp different from the timestamp of the previous one
+    LOG(DEBUG) << "destmap[pid].size() = " << destmap[dpid].size();
+    if (destmap[dpid].size() == 0 || etime != std::abs(latestTimeStamp)) {
+      LOG(DEBUG) << "adding new value";
+      destmap[dpid].push_back(val.payload_pt1);
+      latestTimeStamp = etime;
+    }
+  }
+}
+
+//______________________________________________________________________
+
+template <>
+void DCSProcessor::checkFlagsAndFill(const std::pair<DPID, DPVAL>& dpcom, int64_t& latestTimeStamp, std::unordered_map<DPID, DCSProcessor::DQStrings>& destmap);
+
+//______________________________________________________________________
+
+template <>
+void DCSProcessor::checkFlagsAndFill(const std::pair<DPID, DPVAL>& dpcom, int64_t& latestTimeStamp, std::unordered_map<DPID, DCSProcessor::DQBinaries>& destmap);
+
+//______________________________________________________________________
+
+template <typename T>
+void DCSProcessor::doSimpleMovingAverage(int nelements, std::deque<T>& vect, float& avg, bool& isSMA)
+{
+
+  // Do simple moving average on vector of type T
+
+  if (vect.size() <= nelements) {
+    //avg = std::accumulate(vect.begin(), vect.end(), 0.0) / vect.size();
+    avg = (avg * (vect.size() - 1) + vect[vect.size() - 1]) / vect.size();
```

Comment:
use ``vect.back()`` instead of ``vect[vect.size() - 1]``.

Reply:
Done, thanks